streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -886,6 +944,145 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
         }
     }
 
+  /**
+   * More powerful version of `ZStream#groupByKey`
+   */
+  def groupBy[R1 <: R, E1 >: E, K, V](
+    f: A => ZIO[R1, E1, (K, V)],
+    buffer: Int = 16
+  ): ZStream[R1, E1, (K, Stream[E1, V])] =
+    ZStream.unwrapManaged {
+      for {
+        decider <- Promise.make[Nothing, (K, V) => UIO[Int => Boolean]].toManaged_
+        out     <- Queue.bounded[Take[E1, (K, Stream[E1, V])]](buffer).toManaged(_.shutdown)
+        emit    <- Ref.make[Boolean](true).toManaged_
+        offer    = { (a: Take[E1, (K, Stream[E1, V])]) =>
+           out.offer(a)
+            .catchAllCause {
+              case c if (c.interrupted) => emit.set(false).unit
+            }
+        }
+        ref <- Ref.make[Map[K, Int]](Map()).toManaged_
+        add <- {
+          self
+                .mapM(f)
+                .toQueuesBalanced0(
+                  buffer, { kv: (K, V) =>
+                    decider.await.flatMap(_.tupled(kv))
+                  },
+                  offer(_)
+                )
+        }
+        _ <- decider.succeed {
+              case (k, _) =>
+                ref.get.map(_.get(k)).flatMap {
+                  case Some(idx) => ZIO.succeed(_ == idx)
+                  case None =>
+                    emit.get.flatMap {
+                      case true =>
+                        add.flatMap {
+                          case (idx, q) =>
+                            (ref.update(_ + (k -> idx)) *>
+                              offer(Take.Value(k -> ZStream.fromQueue0(q).unTake.map(_._2)))).as(_ == idx)
+                        }
+                      case false => ZIO.succeed(_ => false)
+                    }
+                }
+            }.toManaged_
+      } yield ZStream.fromQueue0(out).unTake
+    }
+
+  /**
+   * Group a stream using a function. This returns a stream of stream with each stream corresponding
+   * to a key value that has occurred. The driver stream will wait for all child streams to finish.
+   * Note that both the number of active keys and the number of buffered elements per key are limited.
+   * Prefer to use this with `ZStream#flatMapParSema` in order to prevent deadlocks.
+   */
+  def groupByKey[K](f: A => K, buffer: Int = 16): ZStream[R, E, (K, Stream[E, A])] =
+    self.groupBy(a => ZIO.succeed((f(a), a)), buffer)
+
+  /**
+   * Split a stream into n streams using a hash function.
+   * The driver streamer will only ever advance of the maximumLag values before the
+   * slowest downstream stream.
+   * That means that the producer may backpressure easily if the partitions are very uneven.
+   * If a stream finishes the keys will be rebalanced in a consistent way.
+   * This returns a stream and an unsubsribe function that can be used to remove a stream manually.
+   * In that case that stream will receive no new elements and run to completion. Manually unsubscribing
+   * and waiting for the stream to finish ensures that no elements are lost.
+   */
+  def hashPartition(
```

Comment:
I still need to do quite a bit more testing here and must update the docstring before it gets merged.
I added this here for now mainly to get some discussion started on whether we want something like this.
It might make sense to rip this out and do this in a seperate pr though

Reply:
removed for now